hazelcast/src/main/java/com/hazelcast/query/PagingPredicate.java

```diff
@@ -182,7 +182,7 @@ private void setInnerPredicate(Predicate<K, V> predicate) {
 
         Set<QueryableEntry<K, V>> set = ((IndexAwarePredicate<K, V>) predicate).filter(queryContext);
         if (set == null || set.isEmpty()) {
-            return null;
+            return set;
```

Comment:
Since `IndexAwarePredicate#filter` is public API, we should also update its javadoc for `null` return value semantics.

Reply:
Looks like we have a problem here: we are not really supporting `null` results everywhere, for instance `OrPredicate` checks for `null`s, while `AndPredicate` doesn't. To me it feels like it's better to state we are not expecting `null` to be returned from `IndexAwarePredicate.filter` and calling `filter` on the index-aware predicate returned `false` from `IndexAwarePredicate.isIndexed` is an undefined behavior.

But on other hand, these `null`s are apparently also used to indicate a failure state when an index is already destroyed (for instance, during the map destruction), so we can't just state we are not supporting null results and we can't state we are supporting them fully :) Probably this null handling stuff deserves a separate task, wdyt?

Reply:
Sounds good!